### ISSUE TITLE
Double-clicking on a word with a kanji root and hiragana after will highlight the whole word (only).

### ISSUE THREAD

**Issue by r12a** [w3c/character_phrase_tests#30]
[test](https://w3c.github.io/i18n-tests/exploratory/vanilla?text=%E5%A4%A9%E6%B0%97%E4%BA%88%E5%A0%B1%E3%81%AF%E6%99%B4%E3%82%8C%E3%81%A7%E3%81%99%E3%80%82%3Cbr%3E%E7%9A%87%E5%B1%85%E5%91%A8%E8%BE%BA%E3%82%92%E6%AD%A9%E3%81%8D%E3%81%BE%E3%81%97%E3%81%9F%E3%80%82&lang=ja&fontSize=60&width=730&height=216)

Double-click on 晴れ and 歩きました.  Test passes if all those characters and only those characters are highlighted.

- Gecko: ✘ Firefox/75.0 Mac OS X 10.14.6 Highlights either the kanji character or the following hiragana (and any further hiragana following that).
- Blink: ✓✘ Chrome/79.0.3945.117 Mac OS X 10.14.6 Recognises 晴れ but not 歩きました. As per hiragana tests, it tends to highlight only 2 characters at a time.
- Webkit: ✓✘ Safari/605.1.15 Mac OS X 10.14.6 Same as Chrome.

**Comment by xfq** [comment]
I tried double-clicking 歩きました on Chrome 81 and Safari 13.1 (on macOS 10.15.4), but both of them only highlight 歩き without ました.

This *might* make sense because 歩き is the 連用形 (conjunctive form) of 歩く and ました is  just an inflectional suffix, but I'm not sure what the right behavior is and where it is specified.

**Comment by xfq** [comment]
It doesn't seem to be specified in [UAX #29](https://www.unicode.org/reports/tr29/#Word_Boundaries).

**Comment by himorin** [comment]
For Japanese text, 歩きました is correct and should be divided as 歩き まし た by Japanese morphological analysis.
You can try MeCab (one of implementation of Japanese morphological analysis) online at e.g. http://www4414uj.sakura.ne.jp/Yasanichi1/unicheck/

**Comment by xfq** [comment]
In Chrome and Safari it's `歩き ま した` instead of `歩き まし た`, so I guess it's still not correct, because it splits the words ます into two words (ま and the past tense of する, that is, した), instead of regarding まし as the 連用形 of ます, as in MeCab.

**Comment by sideshowbarker** [comment]
> In Chrome and Safari it's `歩き ま した` instead of `歩き まし た`

Yeah, in my macOS environment, it’s segmented into `歩き ま した` in every macOS app I’ve tried it in except Firefox (which instead just seems to do simple splitting at kanji/kana/rōmaji boundaries).

I think macOS has ICU-based grapheme/word segmentation & selection support built into the OS — because the Japanese segmentation/selection behavior is consistent across all macOS apps I’ve tried: TextEdit, Stickies, Notes, Terminal, Safari, and Chrome.

The only exception on macOS is Firefox — and a Firefox engineer tells me that’s because Firefox doesn’t use the built-in macOS platform APIs for it. See https://bugzil.la/345823.

**Comment by r12a** [comment]
This morphological segmentation does indeed stem from ICU. See the [ICU demo page](https://icu4c-demos-7hxm2n5zgq-uc.a.run.app/icu-bin/icusegments#1/en).

I guess the key question, for me, is **what the average Japanese user expects to happen: morphological or whole word selection**?  I understand the academic justification for the former, but does it match user expectations, do they prefer to be able to select the word root (including incidental hiragana, as in 歩き),or do they get frustrated that they have to constantly extend the selection to get what they want, ie. the whole word?

Anyone got a feeling for that? (Or preferably some evidence.)

**Comment by r12a** [comment]
Actually, i think i should raise this as a formal question in the jlreq issue list.  Please reply there.

**Comment by xfq** [comment]
I just wrote [another test](https://w3c.github.io/i18n-tests/exploratory/vanilla/?text=%E6%B5%B7%E5%B2%B8%E3%81%AB%E6%B2%BF%E3%81%A3%E3%81%9F%E5%8C%BA%E5%9F%9F&lang=ja&fontSize=60&width=730&height=216), and the result seems to be less ideal than the test in the OP:

* In Firefox it's: 海岸/に/沿/った/区域
* In Chrome and Safari it's: 海岸/に/沿/っ/た/区域

IMHO the correct behavior should be either 海岸/に/沿っ/た/区域 or 海岸/に/沿った/区域.

**Comment by xfq** [comment]
Here's [another test](https://w3c.github.io/i18n-tests/exploratory/vanilla/?text=%E6%B5%B7%E5%B2%B8%E3%81%AB%E6%B2%BF%E3%81%A3%E3%81%9F%E5%8C%BA%E5%9F%9F&lang=ja&fontSize=60&width=730&height=216) that works in Firefox but not in Chrome/Safari:

* In Firefox it's: 目上/の/方
* In Chrome and Safari it's: 目/上の/方

The reason might be that Firefox only distinguishes between Kanji and Kana, while the dictionary used by Chrome and Safari does not include the word 目上 (which is a common word, so it may also be an issue of the word boundary analysis algorithm).
